File: OsmAnd-java/src/main/java/net/osmand/binary/Abbreviations.java
package net.osmand.binary;

import net.osmand.search.core.SearchPhrase;
import net.osmand.util.SearchAlgorithms;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.ConcurrentHashMap;


public class Abbreviations {

    private Abbreviations() {
    }

    private static final Map<String, Dictionary> DICTIONARIES = new ConcurrentHashMap<>();

	private static Dictionary dictionary(String locale) {
		Objects.requireNonNull(locale, "locale");
		return DICTIONARIES.computeIfAbsent(locale, key -> new Dictionary(SearchVariantRules.forLocale(key)));
	}

	private static final class Dictionary {
		final Map<String, String> abbreviations = new HashMap<>();
		final Map<String, String> searchAbbreviations = new HashMap<>();
		final Set<String> buildingAbbreviations = new TreeSet<>();
		final Set<String> conjunctions = new TreeSet<>();
		final Set<String> commonSkipOtherCnt = new TreeSet<>();

		Dictionary(SearchVariantRules rules) {
			for (SearchVariantRules.Entry entry : rules.indexEntries()) {
				abbreviations.put(entry.key, entry.value);
				commonSkipOtherCnt.add(entry.key);
				commonSkipOtherCnt.add(entry.value.toLowerCase(Locale.ROOT));
			}
			searchAbbreviations.putAll(abbreviations);
			for (SearchVariantRules.Entry entry : rules.queryEntries()) {
				switch (entry.kind) {
					case "search" -> searchAbbreviations.put(entry.key, entry.value);
					case "building" -> buildingAbbreviations.add(entry.key);
					case "conjunction" -> {
						conjunctions.add(entry.key);
						commonSkipOtherCnt.add(entry.key);
					}
				}
			}
		}
	}

	public static boolean likelyPartOfRef(String word, Set<String> wordSplit) {
		int limit = 2;
		int letters = SearchAlgorithms.letters(word, limit + 1);
		if (letters < limit || (letters == limit && SearchAlgorithms.startsWithDigit(word))) {
			return true;
		}
		for (String s : wordSplit) {
			letters = SearchAlgorithms.letters(s, limit + 1);
			if (!(letters < limit || (letters == limit && SearchAlgorithms.startsWithDigit(s)))) {
				return false;
			}
		}
		return true;
	}
	
	// search v-2
	public static boolean likelyPartOfBuilding(String word, Set<String> wordSplit, String locale) {
		Dictionary rules = dictionary(locale);
		boolean bldNum = (SearchAlgorithms.isNumber2Letters(word) || word.length() == 1
				|| rules.buildingAbbreviations.contains(word));
		if (bldNum) {
			return true;
		}
		if (wordSplit != null) {
			// recursion for 2bis
			for (String w : wordSplit) {
				boolean likely = likelyPartOfBuilding(w, null, locale);
				if (!likely) {
					return false;
				}
			}
			return true;
		}
		return false;
	}
    
    
    // search-v2
    public static Map<String, String> getSearchabbreviations(String locale) {
		return dictionary(locale).searchAbbreviations;
	}

	public record QueryForm(SearchVariantRules.Variant rule, String word) {
	}

	/** Only one-word replacements can be matched against one name-index atom. */
	public static List<QueryForm> getQueryForms(String word, String locale) {
		Objects.requireNonNull(locale, "locale");
		List<QueryForm> forms = new ArrayList<>();
		for (SearchVariantRules.Variant rule : SearchVariantRules.forLocale(locale).query()) {
			String replacement = rule.apply(word);
			if (replacement != null) {
				List<String> words = SearchAlgorithms.splitAndNormalize(replacement, false);
				if (words.size() == 1 && !words.get(0).equals(word)) {
					forms.add(new QueryForm(rule, words.get(0)));
				}
			}
		}
		return forms;
	}
    
    // search-v2
	public static boolean isCommonSkipOtherCnt(String lowerCase, String locale) {
		return dictionary(locale).commonSkipOtherCnt.contains(lowerCase);
	}

    // Indexing data
    public static String replaceAll(String phrase, String locale) {
        Map<String, String> abbreviations = dictionary(locale).abbreviations;
        String[] words = phrase.split(SearchPhrase.DELIMITER);
        StringBuilder r = new StringBuilder();
        boolean changed = false;
        for (String w : words) {
            if (r.length() > 0) {
                r.append(SearchPhrase.DELIMITER);
            }
            String abbrRes = abbreviations.get(w.toLowerCase(Locale.ROOT));
            if (abbrRes == null) {
                r.append(w);
            } else {
                changed = true;
                r.append(abbrRes);
            }
        }
        return changed ? r.toString() : phrase;
    }
    
	// search-v1
    public static Map<String, String> getAbbreviations(String locale) {
		return dictionary(locale).abbreviations;
	}

	// search v-1
	public static String replace(String word, String locale) {
		String value = dictionary(locale).abbreviations.get(word.toLowerCase(Locale.ROOT));
        return value != null ? value : word;
    }
    
    // search-v1
	public static boolean isConjunction(String lowerCase, String locale) {
		return dictionary(locale).conjunctions.contains(lowerCase);
	}
	
    
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/binary/Abbreviations.kt
package net.osmand.shared.binary

import net.osmand.shared.util.KSearchAlgorithms

/**
 * A copy of `Abbreviations` in OsmAnd-java: the short forms of street words and directions, the
 * articles and conjunctions a search skips, and the words that go with a house number.
 *
 * `replaceAll`, which only the map creator calls, is not copied.
 */
object Abbreviations {

	private val abbreviations = HashMap<String, String>()

	// 2nd version search abbrevations for spatial search
	private val searchAbbreviations = HashMap<String, String>()

	// set of words to check for buidlings
	private val buildingAbbreviations = HashMap<String, String>()
	private val conjunctions = HashSet<String>()

	private val commonSkipOtherCnt = HashSet<String>()

	private fun addDirectionWord(key: String, full: String) {
		abbreviations[key] = full
		commonSkipOtherCnt.add(key)
		commonSkipOtherCnt.add(full.lowercase())
	}

	private fun addStreetStatus(key: String, full: String) {
		abbreviations[key] = full
		commonSkipOtherCnt.add(key)
		commonSkipOtherCnt.add(full.lowercase())
	}

	private fun addConjunction(key: String) {
		conjunctions.add(key)
		commonSkipOtherCnt.add(key)
	}

	init {
		// articles
		addConjunction("the")
		addConjunction("de")
		addConjunction("du")
		addConjunction("der")
		addConjunction("den")
		addConjunction("die")
		addConjunction("das")
		addConjunction("la")
		addConjunction("le")
		addConjunction("el")
		addConjunction("il")
		addConjunction("of")

		// and
		addConjunction("and")
		addConjunction("und")
		addConjunction("en")
		addConjunction("et")
		addConjunction("y")
		addConjunction("и")

		// direction
		addDirectionWord("e", "East")
		addDirectionWord("w", "West")
		addDirectionWord("s", "South")
		addDirectionWord("n", "North")
		addDirectionWord("sw", "Southwest")
		addDirectionWord("se", "Southeast")
		addDirectionWord("nw", "Northwest")
		addDirectionWord("ne", "Northeast")

		// street status
		addStreetStatus("ln", "Lane")
		addStreetStatus("dr", "Drive")
		addStreetStatus("rd", "Road")
		addStreetStatus("av", "Avenue")
		addStreetStatus("st", "Street") // 2 values could be saint
		addStreetStatus("hwy", "Highway")
		addStreetStatus("blvd", "Boulevard")

		searchAbbreviations.putAll(abbreviations)
		searchAbbreviations["ave"] = "Avenue" // extra
		searchAbbreviations["st"] = "Street Saint" // 2 values could be saint
		// duplicates - synonyms and not abbrevations actually
		searchAbbreviations["о"] = "Остров"
		searchAbbreviations["остров"] = "о."
		searchAbbreviations["1st"] = "First"
		searchAbbreviations["2nd"] = "Second"
		searchAbbreviations["3rd"] = "Third"
		searchAbbreviations["first"] = "1st"
		searchAbbreviations["second"] = "2nd"
		searchAbbreviations["third"] = "3rd"
		searchAbbreviations["fourth"] = "4th"
		searchAbbreviations["fifth"] = "5th"
		searchAbbreviations["sixth"] = "6th"
		searchAbbreviations["seventh"] = "7th"

		// common housenumber additions
		// french
		buildingAbbreviations["bis"] = "Bis"
		buildingAbbreviations["ter"] = "Ter"
		buildingAbbreviations["quater"] = "Quater"
		// american
		buildingAbbreviations["bldg"] = "Building"
		buildingAbbreviations["ste"] = "Suite"
		buildingAbbreviations["unt"] = "Unit"
		buildingAbbreviations["apt"] = "Apartment"
		buildingAbbreviations["fl"] = "Floor"
		buildingAbbreviations["flr"] = "Floor"
		buildingAbbreviations["bsmt"] = "Basement"
	}

	fun likelyPartOfRef(word: String, wordSplit: Set<String>): Boolean {
		val limit = 2
		var letters = KSearchAlgorithms.letters(word, limit + 1)
		if (letters < limit || (letters == limit && KSearchAlgorithms.startsWithDigit(word))) {
			return true
		}
		for (s in wordSplit) {
			letters = KSearchAlgorithms.letters(s, limit + 1)
			if (!(letters < limit || (letters == limit && KSearchAlgorithms.startsWithDigit(s)))) {
				return false
			}
		}
		return true
	}

	// search v-2
	fun likelyPartOfBuilding(word: String, wordSplit: Set<String>?): Boolean {
		val bldNum = KSearchAlgorithms.isNumber2Letters(word) || word.length == 1
				|| buildingAbbreviations.containsKey(word)
		if (bldNum) {
			return true
		}
		if (wordSplit != null) {
			// recursion for 2bis
			for (w in wordSplit) {
				val likely = likelyPartOfBuilding(w, null)
				if (!likely) {
					return false
				}
			}
			return true
		}
		return false
	}

	// search-v2
	fun getSearchabbreviations(): Map<String, String> = searchAbbreviations

	// search-v2
	fun isCommonSkipOtherCnt(lowerCase: String): Boolean = commonSkipOtherCnt.contains(lowerCase)

	// search-v1
	fun getAbbreviations(): Map<String, String> = abbreviations

	// search v-1
	fun replace(word: String): String = abbreviations[word.lowercase()] ?: word

	// search-v1
	fun isConjunction(lowerCase: String): Boolean = conjunctions.contains(lowerCase)
}
